--- a/net/minecraft/world/entity/animal/bee/Bee.java
+++ b/net/minecraft/world/entity/animal/bee/Bee.java
@@ -198,9 +_,18 @@
 
     @Override
     protected void addAdditionalSaveData(final ValueOutput output) {
+        // CraftBukkit start - selectively save data
+        this.addAdditionalSaveData(output, true);
+    }
+
+    @Override
+    public void addAdditionalSaveData(final ValueOutput output, final boolean saveAll) {
+        // CraftBukkit end
         super.addAdditionalSaveData(output);
+        if (saveAll) { // Paper
         output.storeNullable("hive_pos", BlockPos.CODEC, this.hivePos);
         output.storeNullable("flower_pos", BlockPos.CODEC, this.savedFlowerPos);
+        } // Paper
         output.putBoolean("HasNectar", this.hasNectar());
         output.putBoolean("HasStung", this.hasStung());
         output.putInt("TicksSincePollination", this.ticksWithoutNectarSinceExitingHive);
@@ -238,7 +_,7 @@
                 }
 
                 if (poisonTime > 0) {
-                    livingTarget.addEffect(new MobEffectInstance(MobEffects.POISON, poisonTime * 20, 0), this);
+                    livingTarget.addEffect(new MobEffectInstance(MobEffects.POISON, poisonTime * 20, 0), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
                 }
             }
 
@@ -476,7 +_,11 @@
         if (this.hivePos == null) {
             return null;
         } else {
-            return this.isTooFarAway(this.hivePos) ? null : this.level().getBlockEntity(this.hivePos, BlockEntityType.BEEHIVE).orElse(null);
+            // Paper start - move over logic to accommodate isTooFarAway with chunk load check
+            return this.isTooFarAway(this.hivePos) || this.level().getChunkIfLoadedImmediately(this.hivePos.getX() >> 4, this.hivePos.getZ() >> 4) == null
+                ? null
+                : this.level().getBlockEntity(this.hivePos, BlockEntityType.BEEHIVE).orElse(null);
+            // Paper end
         }
     }
 
@@ -508,7 +_,7 @@
         return this.getFlag(FLAG_ROLL);
     }
 
-    public void setRolling(final boolean rolling) {
+    public void setRolling(boolean rolling) { // Paper - Rolling override
         this.setFlag(FLAG_ROLL, rolling);
     }
 
@@ -569,7 +_,7 @@
             if (effect != null) {
                 this.usePlayerItem(player, hand, heldItem);
                 if (!this.level().isClientSide()) {
-                    this.addEffect(effect);
+                    this.addEffect(effect, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.FOOD); // Paper - Add missing effect cause
                 }
 
                 return InteractionResult.SUCCESS;
@@ -641,8 +_,9 @@
         if (this.isInvulnerableTo(level, source)) {
             return false;
         } else {
+            if (!super.hurtServer(level, source, damage)) return false; // CraftBukkit - Only stop pollinating if entity was damaged
             this.beePollinateGoal.stopPollinating();
-            return super.hurtServer(level, source, damage);
+            return true; // CraftBukkit - Only stop pollinating if entity was damaged
         }
     }
 
@@ -1004,7 +_,7 @@
                             }
                         }
 
-                        if (growState != null) {
+                        if (growState != null && org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(Bee.this, belowPos, growState)) { // CraftBukkit
                             Bee.this.level().levelEvent(LevelEvent.PARTICLES_BEE_GROWTH, belowPos, 15);
                             Bee.this.level().setBlockAndUpdate(belowPos, growState);
                             Bee.this.incrementNumCropsGrownSincePollination();
@@ -1029,7 +_,7 @@
         @Override
         protected void alertOther(final Mob other, final LivingEntity hurtByMob) {
             if (other instanceof Bee && this.mob.hasLineOfSight(hurtByMob)) {
-                other.setTarget(hurtByMob);
+                other.setTarget(hurtByMob, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY); // CraftBukkit - reason
             }
         }
     }
@@ -1195,7 +_,7 @@
                     Bee.this.dropFlower();
                     this.pollinating = false;
                     Bee.this.remainingCooldownBeforeLocatingNewFlower = 200;
-                } else {
+                } else if (Bee.this.savedFlowerPos != null) { // Paper - add null check since API can manipulate this
                     Vec3 flowerPos = Vec3.atBottomCenterOf(Bee.this.savedFlowerPos).add(0.0, 0.6F, 0.0);
                     if (flowerPos.distanceTo(Bee.this.position()) > 1.0) {
                         this.hoverPos = flowerPos;
