--- a/net/minecraft/world/level/block/PointedDripstoneBlock.java
+++ b/net/minecraft/world/level/block/PointedDripstoneBlock.java
@@ -144,6 +_,11 @@
                 && projectile.mayBreak(serverLevel)
                 && projectile instanceof ThrownTrident
                 && projectile.getDeltaMovement().length() > 0.6) {
+                // CraftBukkit start
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(projectile, blockPos, state.getFluidState().createLegacyBlock())) { // Paper - fix wrong block state
+                    return;
+                }
+                // CraftBukkit end
                 level.destroyBlock(blockPos, true);
             }
         }
@@ -210,10 +_,11 @@
                         if (stalactiteTipPos != null) {
                             if (fluidInfo.get().sourceState.is(Blocks.MUD) && fluid == Fluids.WATER) {
                                 BlockState newState = Blocks.CLAY.defaultBlockState();
-                                level.setBlockAndUpdate(fluidInfo.get().pos, newState);
+                                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, fluidInfo.get().pos, newState, Block.UPDATE_ALL)) { // Paper - Call BlockFormEvent
                                 Block.pushEntitiesUp(fluidInfo.get().sourceState, newState, level, fluidInfo.get().pos);
                                 level.gameEvent(GameEvent.BLOCK_CHANGE, fluidInfo.get().pos, GameEvent.Context.of(newState));
                                 level.levelEvent(LevelEvent.DRIPSTONE_DRIP, stalactiteTipPos, 0);
+                                } // Paper - Call BlockFormEvent
                             } else {
                                 BlockPos cauldronPos = findFillableCauldronBelowStalactiteTip(level, stalactiteTipPos, fluid);
                                 if (cauldronPos != null) {
